design-proposals/crd-based-management.md

```diff
@@ -0,0 +1,417 @@
+ mechanism is currently being replaced by an installation operator which is also based on
+ CRDs.
+
+Current CLI options provide low level access to API operations (i.e., CRUD on object) and
+ consequently relatively low value. In addition, the current API (both `gwctl` and `cl-adm`)
+ do not address roles and permissions: `gwctl` users are either authenticated and authorized
+ to the entire API or not at all; `cl-adm` files are created locally, relaying on file system
+ attributes for protection. The two CLIs are also only partially consistent in UX. Transitioning
+ to CRDs and `kubectl` should address both consistency and security using existing k8s support.
+
+## Goals
+
+- Provide full life cycle management of ClusterLink configuration, that is comparable with
+ current CLI functionality, but accessible via `kubectl`.
+- Enable use of common k8s tooling and management capabilities.
+
+## Non-Goals
+
+- Provide means for central management (e.g., to facilitate fabric management,
+ automation of common tasks such as Peer and Service sharing across clusters).
+- Support for multitenancy and fabric sharing (i.e., interconnect multiple fabrics)
+- Enable per-namespace/unprivileged deployments
+
+All of the above can be future extensions, so make sure to not block them?
+
+### Prior discussion and links
+
+These topics and their related aspects have been discussed, at a high level, in the following issues:
+
+- [25 - CLI harmonization](https://github.com/clusterlink-net/clusterlink/issues/25)
+- [28 - K8s Central Management](https://github.com/clusterlink-net/clusterlink/issues/28)
+- [38 - gwctl output](https://github.com/clusterlink-net/clusterlink/issues/38)
+- [67 - CLI parameters](https://github.com/clusterlink-net/clusterlink/issues/67)
+- [68 - Watch Services and Pods](https://github.com/clusterlink-net/clusterlink/issues/68)
+- [69 - Control plane persistence](https://github.com/clusterlink-net/clusterlink/issues/69)
+- [77 - Privileged deployments](https://github.com/clusterlink-net/clusterlink/issues/77)
+- [79 - Async control plane](https://github.com/clusterlink-net/clusterlink/issues/79)
+- [133 - Use of k8s controller-runtime](https://github.com/clusterlink-net/clusterlink/issues/133)
+
+## User Stories
+
+### Users
+
+We envisage multiple user roles involved in the management of a ClusterLink deployment.
+
+- **Fabric administrator** has control over the entire ClusterLink fabric. They are responsible
+ for creating the fabric's root of trust, controlling which peers are part of the fabric, etc.
+ This role is mostly not in scope for this design document.
+- **Site administrator** has control over a single cluster in the fabric and manages the
+ installation and operation of the local ClusterLink deployment. Site administrators can
+ add remote Peers to their cluster and may define admin level network policies.
+- **Application owners and developers** manage Services and their access policies. They
```

Comment:
Do we support scenarios where fabric admin, site admin and application owners belong to the same entity? In this case, Clusterlink control plane and dataplane would be deployed on a single namespace which the user has access to. Ofcourse we wouldn't have CRD for that scenario. But, I am just concerned if we remove the API access, how we intend to support such scenarios.

Reply:
I would differentiate between two aspects:
- the CRDs are installed (one time operation) and no further admin level access is required; and
- absolutely no global cluster admin privileges.

The first is on our roadmap, and this design doc "hints" at what could be done to support it (e.g., Peer are defined in the same namespace as the CP and DP). I'm not sure how this would work with a true "no cluster privileges" mode as we need to expose the CP/DP Service outside of the cluster which (IMHO) is a privileged operation. Think of the case of multiple different namespace, each with their own deployment - how would you route from the outside to the correct CL instance? You would need to open multiple ingress points into the cluster and I'm not sure this would be considered "unprivileged". Perhaps we can explore the use case you have in mind in more details. Note that if one time admin access is reasonable (e.g., CRD and CL are deployed) all other management would be per namespace.

The second is currently out of scope, but we might be able to support it by "emulating" CRDs as ConfigMap objects (e.g., with `version`, `type`, `spec.*` and `status.*` fields. You lose the type validation provided by the API for CRDs, but the changes in the controllers may be minimized by adding a layer to convert between ConfigMap and the corresponding Go structure for the CRD and then handing the "CRD instance" for further processing.